[clang] __is_trivially_equality_comparable for types containing lambdas

AMP999 opened the pull request:
Lambdas (closure types) are trivially equality-comparable iff they are
non-capturing, because non-capturing lambdas are convertible to function
pointers: if (lam1 == lam2) compiles, then lam1 and lam2 must have
the same type, and be always-equal, and be empty.

AMP999 committed:
[clang] Rename some misleading names (Non-functional)

AMP999 committed:
[clang] __is_trivially_equality_comparable for types containing lambdas

Lambdas (closure types) are trivially equality-comparable iff they are
non-capturing, because non-capturing lambdas are convertible to function
pointers: if `(lam1 == lam2)` compiles, then `lam1` and `lam2` must have
the same type, and be always-equal, and be empty.

cor3ntin commented on clang/lib/AST/Type.cpp:
```diff
@@ -2663,6 +2663,9 @@ static bool
 HasNonDeletedDefaultedEqualityComparison(const CXXRecordDecl *Decl) {
   if (Decl->isUnion())
     return false;
+  if (Decl->isLambda())
+    return Decl->captures().empty() &&
+           (Decl->getLambdaCaptureDefault() == LCD_None);
```
This pattern is used in a few places, maybe we should factor it out

cor3ntin reviewed (approved):
I think there is an opportunity for a small refactor here, but feel free to land this and we can handle the refactor later as an NFC change if you prefer

AMP999 committed:
[clang] Factor out isCapturelessLambda

This changes the behavior of the call-site in "CodeGenFunction.cpp",
I think for the better. But I couldn't figure out how to test that
codepath.

AMP999 commented:
> I think there is an opportunity for a small refactor here, but feel free to land this and we can handle the refactor later as an NFC change if you prefer

Is this (344455e) what you had in mind? I only found a few places that could use the factored-out function, and in the "CodeGenFunction.cpp" case, I'm not even sure it applies. It looks like that codepath is intended to be hit when we codegen the `operator()` of a lambda with captures under `-fsanitize=null`, but I couldn't see any null-check in the generated code even when the lambda did have captures, so I'm not sure I understand it. That codepath is 
@zygoloid's from 2017 (376c28e). Should I keep that diff, or revert it?

cor3ntin commented:
Yup, this is what i have in mind. Afaik, the CodeGenFunction.cpp change was a bug that you fixed - just going off the comment.

efriedma-quic commented:
Do you need to check if the lambda is templated?  For example, the following doesn't compile:

```
auto GetNonCapturingLambda() { return []<T>(){ return 42; }; }
auto f() { return GetNonCapturingLambda() == GetNonCapturingLambda(); }
```
